rfcs/20231005-hybrid-quantization.md

```diff
@@ -0,0 +1,116 @@
+# RFC: Hybrid quantized ops for weight-only quantization
+
+Status: Review<br/>
+Initial version: 10/05/2023<br/>
+Last updated: 10/05/2023<br/>
+
+## Introduction
+
+This RFC proposes to extend the convolution and dot_general ops to allow
+differing element types for their operands and define their semantics to
+represent weight-only quantization. This work is related to the more general
+topic of [mixed precision](https://github.com/openxla/stablehlo/issues/369), but
+does not aim to address it in its entirety.
+
+Weight-only quantization is a quantization scheme which is often used to reduce
+the size of models that are memory-bound. A weight-only quantized graph accepts
+input in float and weight in quantized type, dequantizes the weight and performs
+float computation. Even though weight-only quantization can already be expressed
+using a combination of uniform_dequantize op and float-only op, we propose to
+extend the semantics of some float-only ops as a representation of weight-only
+quantization because of the following reasons.
+
+* It aligns well with the existing behavior of the corresponding HLO ops which
+allow operand types to differ and upcast operands to the highest-precision type
+in such a case.
+* Being able to express weight-only natively in StableHLO, it does not rely on
+pattern matching by backend to avoid patterns that might result in bad
+performance.
+* Constant and dequantize can unintentionally be folded in downstreams and
+hybrid op removes concern on constant folding. We can also consider inserting
+optimization_barrier in between to prevent constant folding, but this requires
+upstream frameworks to embed optimization_barrier and downstreams to pattern
+match additional optimization_barrier. It is hard to guarantee these
+requirements on use-cases across various frameworks and hardwares.
+
+## Examples
+
+Here are examples of hybrid quantized convolution and dot_general. Hybrid
+quantized ops will get float input(or lhs) and quantized weight(or rhs) as
+operands and output float results.
+
+```mlir
+%conv = "stablehlo.convolution"(%input_f, %weight_q) ... : (tensor<...xf32>, !quant.uniform<i8:f32, scale:zp>) -> tensor<...xf32>>
+```
+
+```mlir
+%dot = "stablehlo.dot_general"(%lhs_f, %rhs_q) ... : (tensor<...xf32>, !quant.uniform<i8:f32, scale:zp>) -> tensor<...xf32>>
+```
+
+## Proposed spec changes
+
+We propose to modify a few constraints on quantized convolution and dot_general
```

Comment:
Modify operation semantics and a a few...

Reply:
Done.